client/notifications/notification-reducer.ts

```diff
@@ -1,60 +1,173 @@
+      return assertUnreachable(notification)
   }
 }
 
-export class NotificationState extends Record({
-  list: List<NotificationRecord>(),
-  ids: Set<string>(),
+function updateReadStatus(
+  state: NotificationState,
+  notificationIds: string[],
+  read: boolean,
+): Map<string, NotificationRecord> {
+  return Map(
+    notificationIds
+      .filter(id => state.idToNotification.has(id))
+      .map(id => [id, toNotificationRecord({ ...state.idToNotification.get(id)!.toJS(), read })]),
+  )
+}
+
+export class ClearRequestRecord extends Record({
+  clearedIdToNotification: Map<string, NotificationRecord>(),
+  clearedNotificationIds: OrderedSet<string>(),
 }) {}
 
+class NotificationBaseState extends Record({
+  idToNotification: Map<string, NotificationRecord>(),
+  notificationIds: OrderedSet<string>(),
+
+  // Cache the values of the clear requests so they can easily be restored if the server returns an
+  // error. Doesn't cache the local notifications.
+  clearRequests: Map<string, ClearRequestRecord>(),
+}) {}
+
+export class NotificationState extends NotificationBaseState {
+  get reversedNotificationIds() {
+    return this.notificationIds.reverse()
+  }
+}
+
 export default keyedReducer(new NotificationState(), {
+  ['@notifications/serverInit'](state, { payload: { notifications } }) {
+    return state
+      .update('idToNotification', m =>
+        m.merge(notifications.map(n => [n.id, toNotificationRecord(n)])),
+      )
+      .update('notificationIds', s => s.union(notifications.map(n => n.id)))
+  },
+
   ['@notifications/add'](state, { payload: { notification } }) {
-    if (state.ids.has(notification.id)) {
+    if (state.idToNotification.has(notification.id)) {
       return state
     }
 
     return state
-      .update('list', l => l.unshift(toNotificationRecord(notification)))
-      .update('ids', s => s.add(notification.id))
+      .update('idToNotification', m => m.set(notification.id, toNotificationRecord(notification)))
+      .update('notificationIds', s => s.add(notification.id))
   },
 
-  ['@notifications/clear'](state) {
-    return state.set('list', List()).set('ids', Set())
+  ['@notifications/clearById'](state, { payload: { notificationId } }) {
+    return state
+      .deleteIn(['idToNotification', notificationId])
+      .deleteIn(['notificationIds', notificationId])
   },
 
-  ['@notifications/markRead'](state) {
-    return state.update('list', l => l.map(n => n.set('unread', false)))
+  ['@notifications/clearBegin'](state, { payload: { reqId, timestamp } }) {
+    const clearedIdToNotification = state.idToNotification.filter(
+      n => timestamp && n.createdAt <= timestamp,
+    )
+    // Preserve the order of the cleared notification IDs set
+    const clearedNotificationIds = state.notificationIds.filter(id =>
+      clearedIdToNotification.has(id),
+    )
+
+    // Apply the clear changes optimistically
+    return state
+      .update('idToNotification', map =>
```

Comment:
think we already went through a round of many places in this PR where `map` was a bad name choice, now it's made its way back in >_<